neps/nep-0508.md

```diff
@@ -0,0 +1,264 @@
+
+## Motivation
+
+Currently, NEAR protocol has four shards. With more partners onboarding, we started seeing that some shards occasionally become over-crowded with respect to total state size and number of transactions. In addition, with state sync and stateless validation, validators will not need to track all shards and validator hardware requirements can be greatly reduced with smaller shard size. With future in-memory tries, it's also important to limit the size of individual shards.
+
+## Specification
+
+### High level assumptions
+
+* Flat storage is enabled.
+* Shard split boundary is predetermined and hardcoded. In other words, necessity of shard splitting is manually decided.
+* For the time being resharding as an event is only going to happen once but we would still like to have the infrastructure in place to handle future resharding events with ease.
+* Merkle Patricia Trie is the underlying data structure for the protocol state.
+* Epoch is at least 6 hrs long for resharding to complete.
+
+### High level requirements
+
+* Resharding must be fast enough so that both state sync and resharding can happen within one epoch.
+* Resharding should work efficiently within the limits of the current hardware requirements for nodes.
+* Potential failures in resharding may require intervention from node operator to recover.
+* No transaction or receipt must be lost during resharding.
+* Resharding must work regardless of number of existing shards.
+* No apps, tools or code should hardcode the number of shards to 4.
+
+### Out of scope
+
+* Dynamic resharding
+  * automatically scheduling resharding based on shard usage/capacity
+  * automatically determining the shard layout
+* Merging shards or boundary adjustments
+* Shard reshuffling
+
+### Required protocol changes
+
+A new protocol version will be introduced specifying the new shard layout which would be picked up by the resharding logic to split the shard.
+
+### Required state changes
+
+* For the duration of the resharding the node will need to maintain a snapshot of the flat state and related columns. As the main database and the snapshot diverge this will cause some extent of storage overhead.
+* For the duration of the epoch before the new shard layout takes effect, the node will need to maintain the state and flat state of shards in the old and new layout at the same time. The State and FlatState columns will grow up to approx 2x the size. The processing overhead should be minimal as the chunks will still be executed only on the parent shards. There will be increased load on the database while applying changes to both the parent and the children shards. 
+
+### Resharding flow
+
+* The new shard layout will be agreed on offline by the protocol team and hardcoded in the reference implementation.
```

Comment:
Given that this is finalized, could we also mention the shard layout here?

Reply:
I believe @walnut-the-cat wanted to keep the NEP more generic and not mention the next shard layout. I don't have any strong opinions here, I'm happy to do either. @walnut-the-cat are you happy for me to add it? What was the reason for not having it here?